sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,813 @@
+kernel arguments.  For example, the [SYCL\_ONEAPI\_accessor\_properties][6]
+extension could be redesigned to use compile-time properties.  Such a redesign
+might look like:
+
+[6]: <https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/accessor_properties/SYCL_ONEAPI_accessor_properties.asciidoc>
+
+```
+namespace sycl {
+
+template <typename dataT,
+          int dimensions,
+          access::mode accessmode,
+          access::target accessTarget,
+          access::placeholder isPlaceholder,
+          typename PropertyListT = ext::oneapi::property_list<>>
+class __attribute__((sycl_special_class)) accessor {/* ... */};
+
+} // namespace sycl
+```
+
+Typical usage would look like this (showing a hypothetical property named
+`foo`):
+
+```
+using sycl;
+using sycl::ext::oneapi;
+
+accessor acc(buf, cgh, property_list{no_alias_v, foo_v<32>});
+```
+
+As before, the header file represents the properties with an internal C++
+attribute, where the initial parameters are the names of the properties and
+the subsequent parameters are the property values.  However, this time the
+attribute decorates one of the member variables.
+
+```
+namespace sycl {
+
+template <typename dataT,
+          int dimensions,
+          access::mode accessmode,
+          access::target accessTarget,
+          access::placeholder isPlaceholder,
+          typename PropertyListT = ext::oneapi::property_list<>>
+class __attribute__((sycl_special_class)) accessor {/* ... */};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename dataT,
+          int dimensions,
+          access::mode accessmode,
+          access::target accessTarget,
+          access::placeholder isPlaceholder,
+          typename ...Props>
+class __attribute__((sycl_special_class)) accessor<dataT,
+                                                   dimensions,
+                                                   accessmode,
+                                                   accessTarget,
+                                                   isPlaceholder,
+                                                   property_list<Props...>> {
+  dataT *ptr
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_kernel_parameter_attributes(
+    Props::meta_name..., Props::meta_value...
+    )]]
+#endif
+  ;
+};
+
+} // namespace sycl
+```
+
+Illustrating this with the previous example:
+
+```
+namespace sycl {
+
+template </* ... */>
+class __attribute__((sycl_special_class)) accessor</* ... */> {
+  dataT *ptr
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_kernel_parameter_attributes(
+    "sycl-no-alias",  // Name of first property
+    "sycl-foo",       // Name of second property
+    nullptr,          // First property has no parameter
+    32                // Value of second property
+    )]]
+#endif
+  ;
+};
+
+} // namespace sycl
+```
+
+As the name implies, this C++ attribute is only used to decorate a member
```

Comment:
I'm open to changing the names of these attributes, but I want to make sure the new names really are better.  My thinking was to name them according to what they do, not according to how they are used.  For example, the name `add_ir_global_variable_attributes` describes what the CFE does with the attribute's parameters.  This naming style is similar to the existing `[[clang::annotate]]`, for example.  By naming the attributes this way, we could use them for other purposes that are unrelated to properties.  In fact, the design document for [device global](https://github.com/intel/llvm/blob/sycl/sycl/doc/DeviceGlobal.md#attributes-attached-to-the-class) does use `add_ir_global_variable_attributes` to add an IR attribute ("sycl-device-global-size"), which does not correspond to a SYCL property.

If we follow your naming proposal, I think we would have these three names:

* `add_ir_global_variable_attributes` -> `global_variable_properties`
* `add_ir_kernel_parameter_attributes` -> `kernel_parameter_properties`
* `add_ir_function_attributes` -> `kernel_function_properties`

The final name is a little less clear.  Would it be something like:

* `add_ir_member_annotation` -> `variable_access_properties`

Reply:
> I'm open to changing the names of these attributes, but I want to make sure the new names really are better. My thinking was to name them according to what they do, not according to how they are used. 

@premanandrao do you have a preference on names here?

>By naming the attributes this way, we could use them for other purposes that are unrelated to properties. In fact, the design document for device global does use add_ir_global_variable_attributes to add an IR attribute ("sycl-device-global-size"), which does not correspond to a SYCL property.

How would we handle these attributes in a generic manner then? How would pass list + required even number of arguments (i.e. property names + values) come into play when you add "sycl-device-global-size" as in that design document?

Reply:
> How would we handle these attributes in a generic manner then? How would pass list + required even number of arguments (i.e. property names + values) come into play when you add "sycl-device-global-size" as in that design document?

I don't understand the concern here.  If you add "sycl-device-global-size" to the parameter list, you also add a parameter for its value.  Therefore, the number of parameters is still even.  If the attribute has a "pass list" and also has "sycl-device-global-size", then you need to add "sycl-device-global-size" to the pass list if you want it to be added to the IR.

Here is an example showing this use case.  This example does not have a "pass list".

```
template <typename T, typename ...Props>
class
#ifdef __SYCL_DEVICE_ONLY__
  [[__sycl_detail__::add_ir_global_variable_attributes(
    "sycl-device-global-size",
    Props::meta_name...,
    sizeof(T),
    Props::meta_value...
    )]]
#endif
  device_global<T, property_list<Props...>> {/*...*/};
```

Reply:
OK. For some reason my brain didn't process that `sizeof(T)` is the value for `"sycl-device-global-size"`. Thanks for explaining. 

In that case I guess the current name makes more sense. I still prefer `add_ir_attributes_global_variable` because when I first read this name I read it as "add LLVM IR for global variable attribute" as opposed to "add IR attributes for global variable". This is however a nit and you can ignore if you prefer current name

Reply:
I can see that.  I'd like to rename them all, though, to be consistent:

* `add_ir_global_variable_attributes` -> `add_ir_attributes_global_variable`
* `add_ir_kernel_parameter_attributes` -> `add_ir_attributes_kernel_parameter`
* `add_ir_function_attributes` -> `add_ir_attributes_function`
* `add_ir_member_annotation` -> `add_ir_annotations_member`

Unless you have objections, I'll make this change.

Reply:
That sounds good to me. Thanks!

Reply:
Done in 9ad46514427b6e99783f955e6092cccaafacdac2